k2/csrc/thread_pool.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+#ifndef K2_CSRC_THREAD_POOL_H_
+#define K2_CSRC_THREAD_POOL_H_
+
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <thread>  // NOLINT
+#include <vector>
+
+#include "k2/csrc/log.h"
+
+namespace k2 {
+
+class ThreadPool {
+ public:
+  using TaskFunc = std::function<void()>;
+
+  /* Create a pool with given number of threads.
+   *
+   * @param [in] num_threads  The number of threads in the pool. If it is <= 0,
+   *                          the number of threads in the pool is set to
+   *                          `std::thread::hardware_concurrency()`.
+   */
+  explicit ThreadPool(int32_t num_threads);
+
+  /* It ask all threads in the pool to exit after processing what they
+   * are currently executing. If there are still tasks in the queue, they
```

Comment:
what do you mean here by saying `If there are still tasks in the queue, they will not be executed`? I was thinking the destructor is just join all threads in it.

Reply:
Mm.  Wouldn't it be safer to have the destructor wait for all tasks to be completed?
I have a feeling not doing so would lead to incorrect behavior in some circumstances.

Reply:
Fixed. The destructor will wait until all submitted tasks are finished.